File: 21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        ListNode* head = new ListNode(0);
        ListNode* dummy = head;

        while(list1 != nullptr && list2 != nullptr){
            if(list1 -> val < list2 -> val){
                dummy -> next = list1;
                list1 = list1 -> next;
            }
            else{
                dummy -> next = list2;
                list2 = list2 -> next;
            }
            dummy = dummy -> next;
        }

        if(list1 == nullptr) dummy -> next = list2;
        else if(list2 == nullptr) dummy -> next = list1;

        return head -> next;
    }
};

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        // Create a dummy node to simplify the code
        val dummy = ListNode(0)
        var current: ListNode? = dummy
        var l1 = list1
        var l2 = list2
        
        // Traverse both lists and compare nodes
        while (l1 != null && l2 != null) {
            if (l1.`val` <= l2.`val`) {
                current?.next = l1
                l1 = l1.next
            } else {
                current?.next = l2
                l2 = l2.next
            }
            current = current?.next
        }
        
        // Attach the remaining nodes from either list
        current?.next = l1 ?: l2
        
        return dummy.next
    }
}